exporter/gcloudpubsubexporter/exporter.go

```diff
@@ -0,0 +1,166 @@
+	//
+	userAgent string
+	config    *Config
+	//
+	client *pubsub.Client
+}
+
+func (*pubsubExporter) Name() string {
+	return name
+}
+
+func (ex *pubsubExporter) Start(ctx context.Context, _ component.Host) error {
+	if ex.client == nil {
+		copts, _ := ex.generateClientOptions()
+		client, _ := pubsub.NewClient(context.Background(), ex.config.ProjectID, copts...)
+		ex.client = client
+	}
+
+	if ex.tracesTopic == nil && ex.tracesTopicName != "" {
+		ex.tracesTopic = ex.client.TopicInProject(ex.tracesTopicName, ex.config.ProjectID)
+		if ex.config.ValidateExistence {
+			tctx, cancel := context.WithTimeout(ctx, ex.config.Timeout)
+			defer cancel()
+			exist, err := ex.tracesTopic.Exists(tctx)
+			if err != nil {
+				return err
+			}
+			if !exist {
+				return fmt.Errorf("trace subscription %s doesn't exist", ex.tracesTopic)
+			}
+		}
+	}
+	if ex.metricsTopic == nil && ex.metricsTopicName != "" {
+		ex.metricsTopic = ex.client.TopicInProject(ex.metricsTopicName, ex.config.ProjectID)
+		if ex.config.ValidateExistence {
+			tctx, cancel := context.WithTimeout(ctx, ex.config.Timeout)
+			defer cancel()
+			exist, err := ex.metricsTopic.Exists(tctx)
+			if err != nil {
+				return err
+			}
+			if !exist {
+				return fmt.Errorf("metric subscription %s doesn't exist", ex.tracesTopic)
+			}
+		}
+	}
+	if ex.logsTopic == nil && ex.logsTopicName != "" {
+		ex.logsTopic = ex.client.TopicInProject(ex.logsTopicName, ex.config.ProjectID)
+		if ex.config.ValidateExistence {
+			tctx, cancel := context.WithTimeout(ctx, ex.config.Timeout)
+			defer cancel()
+			exist, err := ex.logsTopic.Exists(tctx)
+			if err != nil {
+				return err
+			}
+			if !exist {
+				return fmt.Errorf("log subscription %s doesn't exist", ex.tracesTopic)
+			}
+		}
+	}
+	return nil
+}
+
+func (ex *pubsubExporter) Shutdown(context.Context) error {
+	if ex.tracesTopic != nil {
+		ex.tracesTopic.Stop()
+		ex.tracesTopic = nil
+	}
+	if ex.metricsTopic != nil {
+		ex.metricsTopic.Stop()
+		ex.metricsTopic = nil
+	}
+	if ex.logsTopic != nil {
+		ex.logsTopic.Stop()
+		ex.logsTopic = nil
+	}
+	if ex.client != nil {
+		ex.client.Close()
+		ex.client = nil
+	}
+	return nil
+}
+
+func (ex *pubsubExporter) generateClientOptions() ([]option.ClientOption, error) {
+	var copts []option.ClientOption
+	if ex.userAgent != "" {
+		copts = append(copts, option.WithUserAgent(ex.userAgent))
+	}
+	if ex.config.Endpoint != "" {
+		if ex.config.UseInsecure {
+			var dialOpts []grpc.DialOption
+			if ex.userAgent != "" {
+				dialOpts = append(dialOpts, grpc.WithUserAgent(ex.userAgent))
+			}
+			conn, _ := grpc.Dial(ex.config.Endpoint, append(dialOpts, grpc.WithInsecure())...)
+			copts = append(copts, option.WithGRPCConn(conn))
+		} else {
+			copts = append(copts, option.WithEndpoint(ex.config.Endpoint))
+		}
+	}
+	return copts, nil
+}
+
+func (ex *pubsubExporter) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	bytes, _ := td.ToOtlpProtoBytes()
+	message := &pubsub.Message{Data: bytes}
```

Comment:
Do you think we should put a PR in for the `opentelemetry-specification` to document how OTLP interacts with Pub/Sub? While it's straightforward to push the bytes into Data, there could still be different interpretations of how Pub/Sub is supposed to work so we probably want to make sure they're unified across implementations.

Reply:
Thanks @anuraaga. I'll make one comment in reply to your initial review. Thanks, I needed a second eye on this.

- I will add it to the spec. I'll ask in gitter to see if I should do a PR or an OTEP first. I see that Kafka does it the same way as I (sending the Request), so that's good.
- I will split up the PR but will keep this one open in the meantime (I converted it to draft), I want to have someone from Google involved todo a quick scan.
- All fix your concerns.

Reply:
Looking forward to reviewing the spec you draft up, @alexvanboxel! 📘